Cython/Compiler/Code.py

```diff
@@ -1813,27 +1813,15 @@ def normalise_encoding_name(py_string):
             const unsigned int is_unicode : 1;
             const unsigned int is_str : 1;
             const unsigned int intern : 1;
-        } __Pyx_StringTabEntry; /*proto*/
+        } __Pyx_StringTabEntry;
         """ % dict(
             max_length=longest_pystring,
             num_encodings=len(encodings),
         )))
 
-        self.use_utility_code(UtilityCode.load_cached("InitStrings", "StringTools.c"))
-        w = self.parts['pystring_table']
-        w.putln("")
-
-        self.parts['module_state'].putln("PyObject *%s[%s];" % (
-            Naming.stringtab_cname, len(py_strings)))
-
-        self.parts['module_state_clear'].putln("for (int n=0; n<%s; ++n) {" % len(py_strings))
-        self.parts['module_state_clear'].putln("Py_CLEAR(clear_module_state->%s[n]);" %
-            Naming.stringtab_cname)
-        self.parts['module_state_clear'].putln("}")
-        self.parts['module_state_traverse'].putln("for (int n=0; n<%s; ++n) {" % len(py_strings))
-        self.parts['module_state_traverse'].putln("Py_VISIT(traverse_module_state->%s[n]);" %
-            Naming.stringtab_cname)
-        self.parts['module_state_traverse'].putln("}")
+        py_string_count = len(py_strings)
+        self.parts['module_state'].putln(f"PyObject *{Naming.stringtab_cname}[{py_string_count}];")
+        self._generate_module_array_traverse_and_clear(Naming.stringtab_cname, py_string_count)
```

Comment:
I know it's in the old code too, but why are we `Py_VISIT`ing the module strings since they can't be part of reference cycles?

Reply:
[The documentation](https://docs.python.org/3/c-api/typeobj.html#c.PyTypeObject.tp_traverse) of `tp_traverse` says (about their given example):
"""
Note that Py_VISIT() is called only on those members that can participate in reference cycles. Although there is also a `self->key` member, it can only be NULL or a Python string and therefore cannot be part of a reference cycle.

On the other hand, even if you know a member can never be part of a cycle, as a debugging aid you may want to visit it anyway just so the gc module’s `get_referents()` function will include it.
"""
Thus, it's not entirely without use. But it would probably speed up the GC runs a bit if we excluded visits of immutable objects (strings and integers). CPython itself can be pretty rigorous with this, e.g. this is the `tp_traverse` function of a `CodeObject`, ignoring lots of immutable attribute values:
```c
static int
code_traverse(PyCodeObject *co, visitproc visit, void *arg)
{
    Py_VISIT(co->co_consts);
    return 0;
}
```
We could actually avoid both visit and clear for strings and integers. I'll merge this PR (it's a refactoring) and then write a PR for the removal.